logger/logger_types.go

```diff
@@ -2,34 +2,90 @@ package logger
 
 import (
 	"io"
-	"log"
-	"os"
+	"sync"
 )
 
-// Logging struct that holds all user configurable options for the logger
-type Logging struct {
-	Enabled              *bool  `json:"enabled,omitempty"`
-	File                 string `json:"file"`
-	ColourOutput         bool   `json:"colour"`
-	ColourOutputOverride bool   `json:"colourOverride,omitempty"`
-	Level                string `json:"level"`
-	Rotate               bool   `json:"rotate"`
+const timestampFormat = " 02/01/2006 15:04:05 "
+
+const spacer = "|"
+
+type Config struct {
+	Enabled          *bool            `json:"enabled"`
+	AdvancedSettings advancedSettings `json:"advancedSettings"`
+	SubLoggers       []SubLoggers     `json:"subloggers"`
 }
 
-var (
-	debugLogger *log.Logger
-	infoLogger  *log.Logger
-	warnLogger  *log.Logger
-	errorLogger *log.Logger
-	fatalLogger *log.Logger
+type headers struct {
+	Info  string `json:"info"`
+	Warn  string `json:"warn"`
+	Debug string `json:"debug"`
+	Error string `json:"error"`
+}
+
+type advancedSettings struct {
+	Spacer          string  `json:"spacer"`
+	TimeStampFormat string  `json:"timeStampFormat"`
+	Headers         headers `json:"headers"`
+}
 
-	logFileHandle *os.File
+type SubLoggers struct {
+	Name   string `json:"name"`
+	Level  string `json:"level"`
+	Output string `json:"output"`
+}
 
-	logOutput io.Writer
+type Logger struct {
+	Timestamp                                        string
+	InfoHeader, ErrorHeader, DebugHeader, WarnHeader string
+	Spacer                                           string
+}
+
+type subLogger struct {
+	Info, Debug, Warn, Error bool
+	output                   io.Writer
+}
 
-	// LogPath location to store logs in
+type LogEvent struct {
+	data   []byte
+	output io.Writer
+}
+
+type multiWriter struct {
+	writers []io.Writer
+	mu      sync.Mutex
+}
+
+var (
+	logger           = &Logger{}
+	GlobalLogConfig  = &Config{}
+	subSystemLoggers = map[string]subLogger{}
+	eventPool        = &sync.Pool{
+		New: func() interface{} {
+			return &LogEvent{
+				data: make([]byte, 0, 80),
+			}
+		},
+	}
 	LogPath string
+)
+
+const (
+	LogGlobal = "log"
```

Comment:
I feel like this can be renamed to "Global" to minimise duplication re log.LogGlobal

Reply:
The subsystem naming is something i am fixing today i am not happy with it overall the way that they are setup and added  

As you mentioned below the first part just accepting a string can cause confusion on what should be passed 

I am going to rework the way that subsystems are registered and remove the "global" as being log subsystem 

You will end up with a default level/output as part of the main config and then any optional subsystems override these settings

something like
```js
 "logging": {
  "enabled": true,
  "level": "INFO|DEBUG|WARN|ERROR",
  "output": "stdout",
  "advancedSettings": {
   "spacer": " | ",
   "timeStampFormat": " 02/01/2006 15:04:05 ",
   "headers": {
    "info": "[INFO] ",
    "warn": "[WARN] ",
    "debug": "[DEBUG]",
    "error": "[ERROR]"
   }
  },
  "subloggers": [
   {
    "name": "ticker",
    "level": "INFO",
    "output": "file"
   },
   {
    "name": "portfolio",
    "level": "INFO|DEBUG",
    "output": "file"
   }
  ]
 }
```